go/vt/vtgateproxy/discovery.go

```diff
@@ -96,60 +91,87 @@ func RegisterJsonDiscovery() {
-	gate_type string
-	az_id     string
-}
+type hostFilters = map[string]string
 
 // exampleResolver is a
 // Resolver(https://godoc.org/google.golang.org/grpc/resolver#Resolver).
-type resolveJSONGateConfig struct {
-	target          resolver.Target
-	cc              resolver.ClientConn
-	jsonPath        string
-	ticker          *time.Ticker
-	rand            *rand.Rand // safe for concurrent use.
-	num_connections int
-	filters         resolveFilters
+type JSONGateConfigResolver struct {
+	target   resolver.Target
+	cc       resolver.ClientConn
+	jsonPath string
+	ticker   *time.Ticker
+	rand     *rand.Rand // safe for concurrent use.
+	filters  hostFilters
 }
 
-type discoverySlackAZ struct{}
-type discoverySlackType struct{}
+type matchesFilter struct{}
 
-func (r *resolveJSONGateConfig) loadConfig() (*[]resolver.Address, []byte, error) {
-	config := []DiscoveryHost{}
-	fmt.Printf("Loading config %v\n", r.jsonPath)
+func (r *JSONGateConfigResolver) loadConfig() (*[]resolver.Address, []byte, error) {
+	pairs := []map[string]interface{}{}
+	fmt.Printf("Loading config %v config for %v connections\n", r.jsonPath, *numConnectionsInt)
 
 	data, err := os.ReadFile(r.jsonPath)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	err = json.Unmarshal(data, &config)
+	err = json.Unmarshal(data, &pairs)
 	if err != nil {
 		fmt.Printf("parse err: %v\n", err)
 		return nil, nil, err
 	}
 
-	addrs := []resolver.Address{}
-	for _, s := range config {
-		az := attributes.New(discoverySlackAZ{}, s.AZId).WithValue(discoverySlackType{}, s.Type)
+	allAddrs := []resolver.Address{}
+	filteredAddrs := []resolver.Address{}
+	var addrs []resolver.Address
+	for _, pair := range pairs {
+		matchesAll := true
+		for k, v := range r.filters {
+			if pair[k] != v {
+				matchesAll = false
+			}
+		}
 
-		// Filter hosts to this gate type
-		if r.filters.gate_type != "" {
-			if r.filters.gate_type != s.Type {
-				continue
+		if matchesAll {
+			filteredAddrs = append(filteredAddrs, resolver.Address{
+				Addr:               fmt.Sprintf("%s:%s", pair["nebula_address"], pair["grpc"]),
+				BalancerAttributes: attributes.New(matchesFilter{}, "match"),
+			})
+		}
+
+		// Must filter by type
+		t, ok := r.filters["type"]
+		if ok {
+			if pair["type"] == t {
+				// Add matching hosts to registration list
+				allAddrs = append(allAddrs, resolver.Address{
+					Addr:               fmt.Sprintf("%s:%s", pair["nebula_address"], pair["grpc"]),
+					BalancerAttributes: attributes.New(matchesFilter{}, "nomatch"),
+				})
 			}
 		}
+	}
 
-		// Add matching hosts to registration list
-		addrs = append(addrs, resolver.Address{
-			Addr:               fmt.Sprintf("%s:%s", s.NebulaAddress, s.Grpc),
-			BalancerAttributes: az,
-		})
+	fmt.Printf("-----\n")
+	fmt.Printf("filtered: %v\n", filteredAddrs)
+	fmt.Printf("-----\n")
+	fmt.Printf("all: %v\n", allAddrs)
+	fmt.Printf("----\n")
+
+	// Nothing in the filtered list? Get them all
+	if len(filteredAddrs) == 0 {
```

Comment:
Hmm... what if you want 4 connections but there are only 2 in the same AZ. Seems like you would want to make sure to include those 2, but my read of the code below would ignore that and just fall back to the full set.

Implementation-wise I kinda think we could make this all a bit simpler if we did an approach like what I did here https://github.com/slackhq/vitess/pull/41/files#diff-7bddb8b37e422b4f434e7a7e7e55ee318a195c1235ca3284c18d5d11a640f3a4R186

Basically, take the whole json list, and first filter it to do the type matching, since that (as said above) is a real filter.

Then shuffle the whole thing randomly.

Then do another pass where we take the shuffled list and boost the ones that match the AZ affinity to the front of the list.

Then grab the first N.

This way you handle all the cases a bit more elegantly I think. Yes, it's a 2nd pass on the list, but this is small and since we only do this for the first connection to a given target I think it should be fine. If this is actually done each time we dial then we have bigger problems.